File: app/src/main/java/mega/privacy/android/app/listeners/GetAttrUserListener.java
package mega.privacy.android.app.listeners;

import android.content.Context;
import android.content.Intent;

import mega.privacy.android.app.AuthenticityCredentialsActivity;
import mega.privacy.android.app.MegaApplication;
import mega.privacy.android.app.R;
import mega.privacy.android.app.lollipop.FileExplorerActivityLollipop;
import mega.privacy.android.app.lollipop.MyAccountInfo;
import mega.privacy.android.app.lollipop.megachat.ChatActivityLollipop;
import mega.privacy.android.app.lollipop.megachat.GroupChatInfoActivityLollipop;
import mega.privacy.android.app.lollipop.megachat.NodeAttachmentHistoryActivity;
import nz.mega.sdk.MegaAccountDetails;
import nz.mega.sdk.MegaApiJava;
import nz.mega.sdk.MegaError;
import nz.mega.sdk.MegaNode;
import nz.mega.sdk.MegaRequest;
import nz.mega.sdk.MegaUser;

import static mega.privacy.android.app.constants.BroadcastConstants.*;
import static mega.privacy.android.app.listeners.CreateFolderListener.ExtraAction.MY_CHAT_FILES;
import static mega.privacy.android.app.utils.Constants.CHAT_FOLDER;
import static mega.privacy.android.app.utils.ContactUtil.*;
import static mega.privacy.android.app.utils.LogUtil.*;
import static nz.mega.sdk.MegaApiJava.*;

public class GetAttrUserListener extends BaseListener {

    private static final int DAYS_USER_FREE = 30;
    private static final int DAYS_USER_PRO = 90;

    /**
     * Indicates if the request is only to update the DB.
     * If so, the rest of the actions in onRequestFinish() can be ignored.
     */
    private boolean onlyDBUpdate;

    private int holderPosition;

    public GetAttrUserListener(Context context) {
        super(context);
    }

    /**
     * Constructor to init a request for check the USER_ATTR_MY_CHAT_FILES_FOLDER user's attribute
     * and update the DB with the result.
     *
     * @param context      current application context
     * @param onlyDBUpdate true if the purpose of the request is only update the DB, false otherwise
     */
    public GetAttrUserListener(Context context, boolean onlyDBUpdate) {
        super(context);

        this.onlyDBUpdate = onlyDBUpdate;
    }

    /**

     * Constructor to init a request for check the USER_ATTR_AVATAR user's attribute
     * and updates the holder of the adapter.
     *
     * @param context           current application context
     * @param holderPosition    position of the holder to update
     */
    public GetAttrUserListener(Context context, int holderPosition) {
        super(context);

        this.holderPosition = holderPosition;
    }

    /**
     * When calling {@link MegaApiJava#getUserAttribute(int)}, the MegaRequest object won't store
     * node handle, so we can't get user handle from {@code request.getNodeHandle()}, we should
     * use {@code api.getContact(request.getEmail())} to get MegaUser, then get user handle from it.
     */
    @Override
    public void onRequestFinish(MegaApiJava api, MegaRequest request, MegaError e) {
        if (request.getType() != MegaRequest.TYPE_GET_ATTR_USER) return;

        switch (request.getParamType()) {
            case USER_ATTR_MY_CHAT_FILES_FOLDER:
                checkMyChatFilesFolderRequest(api, request, e);
                break;

            case USER_ATTR_FIRSTNAME:
                if (e.getErrorCode() == MegaError.API_OK) {
                    updateFirstName(context, request.getText(), request.getEmail());
                    MegaUser user = api.getContact(request.getEmail());
                    if (user != null) {
                        notifyFirstNameUpdate(context, user.getHandle());
                    }
                }
                break;

            case USER_ATTR_LASTNAME:
                if (e.getErrorCode() == MegaError.API_OK) {
                    updateLastName(context, request.getText(), request.getEmail());
                    MegaUser user = api.getContact(request.getEmail());
                    if (user != null) {
                        notifyLastNameUpdate(context, user.getHandle());
                    }
                }
                break;

            case USER_ATTR_ALIAS:
                if (e.getErrorCode() == MegaError.API_OK) {
                    updateDBNickname(api, context, request.getMegaStringMap());
                } else {
                    logError("Error recovering the alias" + e.getErrorCode());
                }
                break;

            case USER_ATTR_ED25519_PUBLIC_KEY:
                if (context instanceof AuthenticityCredentialsActivity) {
                    ((AuthenticityCredentialsActivity) context).setContactCredentials(request, e);
                }
                break;

            case USER_ATTR_AVATAR:
                if (e.getErrorCode() == MegaError.API_OK) {
                    updateAvatar(request);
                }
                break;

            case MegaApiJava.USER_ATTR_RUBBISH_TIME:
                Intent intent = new Intent(ACTION_UPDATE_RB_SCHEDULER);

                if (e.getErrorCode() == MegaError.API_ENOENT) {
                    MyAccountInfo myAccountInfo = MegaApplication.getInstance().getMyAccountInfo();
                    if (myAccountInfo == null)
                        break;

                    intent.putExtra(DAYS_COUNT, myAccountInfo.getAccountType() == MegaAccountDetails.ACCOUNT_TYPE_FREE
                            ? DAYS_USER_FREE
                            : DAYS_USER_PRO);
                } else {
                    intent.putExtra(DAYS_COUNT, request.getNumber());
                }

                MegaApplication.getInstance().sendBroadcast(intent);
                break;

            case MegaApiJava.USER_ATTR_DISABLE_VERSIONS:
                MegaApplication.setDisableFileVersions(request.getFlag());
                MegaApplication.getInstance().sendBroadcast(new Intent(ACTION_UPDATE_FILE_VERSIONS));
                break;

            case MegaApiJava.USER_ATTR_RICH_PREVIEWS:
                if (e.getErrorCode() == MegaError.API_ENOENT) {
                    logWarning("Attribute USER_ATTR_RICH_PREVIEWS not set");
                }

                if (request.getNumDetails() == 1) {
                    MegaApplication.setShowRichLinkWarning(request.getFlag());
                    MegaApplication.setCounterNotNowRichLinkWarning((int) request.getNumber());
                } else if (request.getNumDetails() == 0) {
                    MegaApplication.setEnabledRichLinks(request.getFlag());
                    MegaApplication.getInstance().sendBroadcast(new Intent(BROADCAST_ACTION_INTENT_RICH_LINK_SETTING_UPDATE));
                }
                break;
        }
    }

    /**
     * Checks if the USER_ATTR_MY_CHAT_FILES_FOLDER user's attribute exists when the request finished:
     * - If not exists but exists a file with the old "My chat files" name, it renames it with the old one.
     * - If the old one neither exists, it launches a request to create it.
     *
     * Updates the DB with the result and, if necessary updates the data in the current Context.
     *
     * @param api       MegaApiJava object
     * @param request   result of the request
     * @param e         MegaError received when the request finished
     */
    private void checkMyChatFilesFolderRequest(MegaApiJava api, MegaRequest request, MegaError e) {
        MegaNode myChatFolderNode = null;
        boolean myChatFolderFound = false;

        if (e.getErrorCode() == MegaError.API_OK) {
            myChatFolderNode = api.getNodeByHandle(request.getNodeHandle());
            if (myChatFolderNode == null) {
                myChatFolderNode = api.getNodeByPath(CHAT_FOLDER, api.getRootNode());
            }
        } else if (e.getErrorCode() == MegaError.API_ENOENT) {
            myChatFolderNode = api.getNodeByPath(CHAT_FOLDER, api.getRootNode());

            if (myChatFolderNode != null && !api.isInRubbish(myChatFolderNode)) {
                String name = context.getString(R.string.my_chat_files_folder);

                if (!myChatFolderNode.getName().equals(name)) {
                    api.renameNode(myChatFolderNode, name, new RenameListener(context, true));
                }
                api.setMyChatFilesFolder(myChatFolderNode.getHandle(), new SetAttrUserListener(context));
            }
        } else {
            logError("Error getting \"My chat files\" folder: " + e.getErrorString());
        }
        if (myChatFolderNode != null && !api.isInRubbish(myChatFolderNode)) {
            dBH.setMyChatFilesFolderHandle(myChatFolderNode.getHandle());
            myChatFolderFound = true;
        } else if (!onlyDBUpdate) {
            api.createFolder(context.getString(R.string.my_chat_files_folder), api.getRootNode(), new CreateFolderListener(context, MY_CHAT_FILES));
        }

        if (onlyDBUpdate) {
            return;
        }

        if (context instanceof FileExplorerActivityLollipop) {
            FileExplorerActivityLollipop fileExplorerActivityLollipop = (FileExplorerActivityLollipop) context;
            if (myChatFolderFound) {
                fileExplorerActivityLollipop.setMyChatFilesFolder(myChatFolderNode);
                fileExplorerActivityLollipop.checkIfFilesExistsInMEGA();
            }
        } else if (context instanceof ChatActivityLollipop) {
            ChatActivityLollipop chatActivityLollipop = (ChatActivityLollipop) context;
            if (myChatFolderFound) {
                chatActivityLollipop.setMyChatFilesFolder(myChatFolderNode);

                if (chatActivityLollipop.isForwardingFromNC()) {
                    chatActivityLollipop.handleStoredData();
                } else {
                    chatActivityLollipop.proceedWithAction();
                }
            }
        } else if (context instanceof NodeAttachmentHistoryActivity) {
            NodeAttachmentHistoryActivity nodeAttachmentHistoryActivity = (NodeAttachmentHistoryActivity) context;
            if (myChatFolderFound) {
                nodeAttachmentHistoryActivity.setMyChatFilesFolder(myChatFolderNode);
                nodeAttachmentHistoryActivity.handleStoredData();
            }
        }
    }

    /**
     * If the avatar was correctly obtained in the request, updates the holder in the adapter with it.
     *
     * @param request   result of the request
     */
    private void updateAvatar(MegaRequest request) {
        if (context instanceof GroupChatInfoActivityLollipop) {
            ((GroupChatInfoActivityLollipop) context).updateParticipantAvatar(holderPosition, request.getEmail());
        }
    }
}

File: app/src/main/java/mega/privacy/android/app/listeners/GetAttrUserListener.kt
package mega.privacy.android.app.listeners

import android.content.Context
import android.content.Intent
import dagger.hilt.android.EntryPointAccessors
import kotlinx.coroutines.launch
import mega.privacy.android.app.MegaApplication
import mega.privacy.android.app.R
import mega.privacy.android.app.constants.BroadcastConstants
import mega.privacy.android.app.di.DatabaseEntryPoint
import mega.privacy.android.app.listeners.CreateFolderListener.ExtraAction
import mega.privacy.android.app.main.FileExplorerActivity
import mega.privacy.android.app.main.megachat.ChatActivity
import mega.privacy.android.app.main.megachat.GroupChatInfoActivity
import mega.privacy.android.app.main.megachat.NodeAttachmentHistoryActivity
import mega.privacy.android.app.utils.Constants
import mega.privacy.android.app.utils.ContactUtil
import mega.privacy.android.data.database.DatabaseHandler
import nz.mega.sdk.MegaAccountDetails
import nz.mega.sdk.MegaApiJava
import nz.mega.sdk.MegaError
import nz.mega.sdk.MegaNode
import nz.mega.sdk.MegaRequest
import nz.mega.sdk.MegaRequestListenerInterface
import timber.log.Timber

/**
 * GetAttrUserListener
 * @property context: Context
 */
class GetAttrUserListener constructor(private val context: Context) : MegaRequestListenerInterface {
    /**
     * Indicates if the request is only to update the DB.
     * If so, the rest of the actions in onRequestFinish() can be ignored.
     */
    private var onlyDBUpdate = false
    private var holderPosition = 0
    private val databaseHandler: DatabaseHandler by lazy { MegaApplication.getInstance().dbH }

    private val databaseEntryPoint: DatabaseEntryPoint by lazy {
        EntryPointAccessors.fromApplication(context.applicationContext)
    }

    /**
     * Constructor to init a request for check the USER_ATTR_MY_CHAT_FILES_FOLDER user's attribute
     * and update the DB with the result.
     *
     * @param context      current application context
     * @param onlyDBUpdate true if the purpose of the request is only update the DB, false otherwise
     */
    constructor(context: Context, onlyDBUpdate: Boolean) : this(context) {
        this.onlyDBUpdate = onlyDBUpdate
    }

    /**
     * Constructor to init a request for check the USER_ATTR_AVATAR user's attribute
     * and updates the holder of the adapter.
     *
     * @param context        current application context
     * @param holderPosition position of the holder to update
     */
    constructor(context: Context, holderPosition: Int) : this(context) {
        this.holderPosition = holderPosition
    }

    /**
     * Callback function for onRequestStart
     *
     * @param api : MegaApiJava
     * @param request : MegaRequest
     */
    override fun onRequestStart(api: MegaApiJava?, request: MegaRequest?) {
        // Do nothing
    }

    /**
     * Callback function for onRequestUpdate
     *
     * @param api : MegaApiJava
     * @param request : MegaRequest
     */
    override fun onRequestUpdate(api: MegaApiJava?, request: MegaRequest?) {
        // Do nothing
    }

    /**
     * When calling [MegaApiJava.getUserAttribute], the MegaRequest object won't store
     * node handle, so we can't get user handle from `request.getNodeHandle()`, we should
     * use `api.getContact(request.getEmail())` to get MegaUser, then get user handle from it.
     *
     * @param api : MegaApiJava
     * @param request : MegaRequest
     * @param e: MegaError
     */
    override fun onRequestFinish(api: MegaApiJava, request: MegaRequest, e: MegaError) =
        with(request) {
            if (type == MegaRequest.TYPE_GET_ATTR_USER) {
                when (paramType) {
                    MegaApiJava.USER_ATTR_MY_CHAT_FILES_FOLDER ->
                        checkMyChatFilesFolderRequest(api, this, e)
                    MegaApiJava.USER_ATTR_FIRSTNAME -> if (e.errorCode == MegaError.API_OK) {
                        databaseEntryPoint.applicationScope().launch {
                            if (!email.isNullOrBlank()) {
                                databaseEntryPoint.localRoomGateway.setContactName(
                                    firstName = text,
                                    mail = email
                                )
                            }
                            api.getContact(email)?.let {
                                ContactUtil.notifyFirstNameUpdate(context, it.handle)
                            }
                        }
                    }
                    MegaApiJava.USER_ATTR_LASTNAME -> if (e.errorCode == MegaError.API_OK) {
                        databaseEntryPoint.applicationScope().launch {
                            if (!email.isNullOrBlank()) {
                                databaseEntryPoint.localRoomGateway.setContactLastName(
                                    lastName = text,
                                    mail = email
                                )
                            }
                            api.getContact(email)?.let {
                                ContactUtil.notifyLastNameUpdate(context, it.handle)
                            }
                        }
                    }
                    MegaApiJava.USER_ATTR_ALIAS -> if (e.errorCode == MegaError.API_OK) {
                        ContactUtil.updateDBNickname(api.contacts, context, megaStringMap)
                    } else {
                        Timber.e("Error recovering the alias %s", e.errorCode)
                    }
                    MegaApiJava.USER_ATTR_AVATAR -> if (e.errorCode == MegaError.API_OK) {
                        (context as? GroupChatInfoActivity)
                            ?.updateParticipantAvatar(holderPosition, email)
                    }
                    MegaApiJava.USER_ATTR_RUBBISH_TIME -> {
                        Intent(BroadcastConstants.ACTION_UPDATE_RB_SCHEDULER).run {
                            if (e.errorCode == MegaError.API_ENOENT) {
                                val daysCount =
                                    if (MegaApplication.getInstance().myAccountInfo.accountType == MegaAccountDetails.ACCOUNT_TYPE_FREE) {
                                        DAYS_USER_FREE
                                    } else {
                                        DAYS_USER_PRO
                                    }
                                putExtra(BroadcastConstants.DAYS_COUNT, daysCount)
                            } else {
                                putExtra(BroadcastConstants.DAYS_COUNT, number)
                            }
                            MegaApplication.getInstance().sendBroadcast(this)
                        }
                    }
                    MegaApiJava.USER_ATTR_RICH_PREVIEWS -> {
                        if (e.errorCode == MegaError.API_ENOENT) {
                            Timber.w("Attribute USER_ATTR_RICH_PREVIEWS not set")
                        }
                        if (numDetails == 1) {
                            MegaApplication.isShowRichLinkWarning = flag
                            MegaApplication.counterNotNowRichLinkWarning = number.toInt()
                        } else if (numDetails == 0) {
                            MegaApplication.isEnabledRichLinks = flag
                            MegaApplication.getInstance()
                                .sendBroadcast(Intent(BroadcastConstants.BROADCAST_ACTION_INTENT_RICH_LINK_SETTING_UPDATE))
                        }
                    }
                }
            }
        }

    /**
     * Callback function for onRequestTemporaryError
     *
     * @param api : MegaApiJava
     * @param request : MegaRequest
     * @param e: MegaError
     */
    override fun onRequestTemporaryError(api: MegaApiJava?, request: MegaRequest?, e: MegaError?) {
        // Do nothing
    }

    /**
     * Checks if the USER_ATTR_MY_CHAT_FILES_FOLDER user's attribute exists when the request finished:
     * - If not exists but exists a file with the old "My chat files" name, it renames it with the old one.
     * - If the old one neither exists, it launches a request to create it.
     *
     *
     * Updates the DB with the result and, if necessary updates the data in the current Context.
     *
     * @param api     MegaApiJava object
     * @param request result of the request
     * @param e       MegaError received when the request finished
     */
    private fun checkMyChatFilesFolderRequest(
        api: MegaApiJava,
        request: MegaRequest,
        e: MegaError,
    ) {
        var myChatFolderNode: MegaNode? = null
        var myChatFolderFound = false
        when (e.errorCode) {
            MegaError.API_OK -> {
                myChatFolderNode = api.getNodeByHandle(request.nodeHandle)?.apply {
                    api.getNodeByPath(Constants.CHAT_FOLDER, api.rootNode)
                }
            }
            MegaError.API_ENOENT -> {
                api.getNodeByPath(Constants.CHAT_FOLDER, api.rootNode)?.let {
                    if (!api.isInRubbish(it)) {
                        val name = context.getString(R.string.my_chat_files_folder)
                        if (it.name != name) {
                            api.renameNode(
                                it, name, RenameListener(
                                    isMyChatFilesFolder = true,
                                    context = context
                                )
                            )
                        }
                        api.setMyChatFilesFolder(it.handle, SetAttrUserListener(context))
                    }
                }
            }
            else -> {
                Timber.e("Error getting \"My chat files\" folder: %s", e.errorString)
            }
        }
        if (myChatFolderNode != null && !api.isInRubbish(myChatFolderNode)) {
            databaseHandler.myChatFilesFolderHandle = myChatFolderNode.handle
            myChatFolderFound = true
        } else if (!onlyDBUpdate) {
            api.createFolder(
                context.getString(R.string.my_chat_files_folder),
                api.rootNode,
                CreateFolderListener(context, ExtraAction.MY_CHAT_FILES)
            )
        }

        if (onlyDBUpdate) {
            return
        }

        with(context) {
            if (myChatFolderFound) {
                (this as? FileExplorerActivity)?.let {
                    it.setMyChatFilesFolder(myChatFolderNode)
                    it.checkIfFilesExistsInMEGA()
                }

                (this as? ChatActivity)?.let {
                    it.setMyChatFilesFolder(myChatFolderNode)
                    if (it.isForwardingFromNC) {
                        it.handleStoredData()
                    } else {
                        it.proceedWithAction()
                    }
                }

                (this as? NodeAttachmentHistoryActivity)?.let {
                    it.setMyChatFilesFolder(myChatFolderNode)
                    it.handleStoredData()
                }
            }
        }
    }

    companion object {
        private const val DAYS_USER_FREE = 30
        private const val DAYS_USER_PRO = 90
    }
}